model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt

```diff
@@ -163,19 +182,79 @@ class IgniteStoreClient(jdbcProperties: Properties? = null, private val inmemory
         return cache.invoke(key, ClientIdProcessor())
     }
 
-    override fun <T> runTransaction(body: () -> T): T {
-        val transactions = ignite.transactions()
-        if (transactions.tx() == null) {
-            transactions.txStart().use { tx ->
-                return pendingChangeMessages.runAndFlush {
-                    val result = body()
-                    tx.commit()
-                    result
+    override fun <T> runWrite(body: () -> T): T {
+        return if (localLocks.canWrite()) {
+            body()
+        } else {
+            localLocks.runWrite {
+                globalReadTransactions.acquireAndRun(maxReadTransactions) { // exclusive access
+                    val transactions = ignite.transactions()
+                    check(transactions.tx() == null) { "Already inside a transaction" }
+                    transactions.txStart(TransactionConcurrency.OPTIMISTIC, TransactionIsolation.READ_COMMITTED).use { tx ->
```

Comment:
```suggestion
                    // `OPTIMISTIC` and `READ_COMMITTED` 
                    // are used instead of the default `PESSIMISTIC` and `REPEATABLE_READ`
                    // because they do not acquire fewer locks when writing.
                    // It is ok to acquire fewer locks, 
                    // because the `globalReadTransactions` 
                    // semaphore ensures no writes can happen in parallel.
                    transactions.txStart(TransactionConcurrency.OPTIMISTIC, TransactionIsolation.READ_COMMITTED).use { tx ->
```